### PR TITLE
Always have compute as the first container in the virt-launcher pod.

### PR DESCRIPTION
**Pull request by fedepaol**
**What this PR does / why we need it**:
When we have container disk volumes, those containers are put at the beginning of the container list of the launcher pod, before the compute container.
This may be a problem when using the sriov mutating webhook, which parses the requested network attachment definitions of the pod and modifies the pod adding the requested resources to the first container. If the first container is not the compute one, we are requesting resources for the wrong container (see https://github.com/openshift/ose-sriov-dp-admission-controller/blob/e65f11b5eb7da1918b439a60d7fcc814719b1433/pkg/webhook/webhook.go#L325).

In this PR we ensure the compute container is always the first one of the list.

**Release note**:

```release-note

Change the virt-launcher pod to always have the computer container as the first of the pod.

```

### CODE REVIEW

**Comment by cynepco3hahue** [comment]
> This may be a problem when using the sriov mutating webhook, which parses the requested network attachment definitions of the pod and modifies the pod adding the requested resources to the first container

Why does it add to the first container, can it check the name of the container? It just too magic, to expect that the container that we need is the first one.

**Comment by fedepaol** [comment]
> > This may be a problem when using the sriov mutating webhook, which parses the requested network attachment definitions of the pod and modifies the pod adding the requested resources to the first container
> 
> Why does it add to the first container, can it check the name of the container? It just too magic, to expect that the container that we need is the first one.

I do agree, the problem here is that the network attachment definition is bound to the pod whereas the device is consumed by a container (and the attachment definition does not contain the name of the container it refers to).
It something we don't have control on since it's going to be shipped as part of the sr-iov operator, so right now we are trying to prevent side effects while trying to understand if we can influence in some way the direction of that admission webhook.

**Comment by cynepco3hahue** [comment]
I see, I was sure that SR-IOV operator something that we also working on.

**Comment by fedepaol** [comment]
> I see, I was sure that SR-IOV operator something that we also working on.

My bad, I should have referred to it as the openshift sr-iov operator. We rely on it for a lot of sr-iov related configuration, my understanding is that it's going to be enabled in openshift 4.3 and as part of it they are going to enable the mutating webhook.

File: pkg/virt-controller/services/template.go [hunk 1]
```diff
@@ -720,25 +718,29 @@ func (t *templateService) RenderLaunchManifest(vmi *v1.VirtualMachineInstance) (
...
+		compute.Env = append(compute.Env, k8sv1.EnvVar{Name: varName, Value: resourceName})
 	}
 
 	if _, ok := vmi.Labels[debugLogs]; ok {
-		container.Env = append(container.Env, k8sv1.EnvVar{Name: ENV_VAR_LIBVIRT_DEBUG_LOGS, Value: "1"})
+		compute.Env = append(compute.Env, k8sv1.EnvVar{Name: ENV_VAR_LIBVIRT_DEBUG_LOGS, Value: "1"})
 	}
 
-	containers = append(containers, container)
+	// Make sure the compute container is always the first since the cni mutating webhook will add the
```
Comment (booxter) [inline pkg/virt-controller/services/template.go]: "which cni mutating webhook?" is a legit question, considering that the hook you refer to is not even under kubevirt organization. this should be explained in the comment with more details.

**Comment by booxter** [comment]
AFAIU there is no reason to bump versions of dependencies in this PR. Are there?

**Comment by fedepaol** [comment]
> AFAIU there is no reason to bump versions of dependencies in this PR. Are there?

Nope & thanks for pointing out. It must have been vscode. I normally skip go.mod / go.sum but they probably sneaked in this time as a result of a git commit --all. Will remove them

**Comment by fedepaol** [comment]
/hold

Changed the timeout to see what happens to sriov lane.

**Comment by fedepaol** [comment]
/test pull-kubevirt-e2e-kind-k8s-sriov-1.14.2

let's see if it was a coincidence

**Comment by fedepaol** [comment]
/test pull-kubevirt-e2e-kind-k8s-sriov-1.14.2
one more time

**Comment by fedepaol** [comment]
Just realized this was on hold
/hold cancel

**Comment by rmohr** [comment]
@fedepaol it still contains the timeout change, right?

**Comment by fedepaol** [comment]
> @fedepaol it still contains the timeout change, right?

ops, you are right!

**Comment by fedepaol** [comment]
@rmohr removed, thanks!

**Comment by rmohr** [comment]
@fedepaol want to add a release note?

**Comment by fedepaol** [comment]
> @fedepaol want to add a release note?

@rmohr done!

**Pull request merged**